shanoir-ng-front/src/app/shared/mass-download/mass-download.service.ts

```diff
@@ -0,0 +1,492 @@
+                    }).catch(reason => {
+                        task.message = 'download error : ' + reason;
+                        this.notificationService.pushLocalTask(task);
+                    }).finally(() => {
+                        releaseQueue();
+                    });
+                } catch (error) {
+                    releaseQueue();
+                    throw error;
+                }
+            });
+        }).catch(error => { /* the user clicked 'cancel' in the choose directory window */ });
+    }
+
+    private _downloadFromReport(report: Report, task: Task, parentHandle: FileSystemDirectoryHandle) {
+        if (!report) throw new Error('report can\'t be null !');
+        const noSuccessIds: number[] = Object.keys(report.list).filter(key => report.list[key].status != 'SUCCESS').map(key => parseInt(key));
+        this._downloadByIds(noSuccessIds, report.format, report.nbQueues, task, report, parentHandle);
+    }
+
+    private handleEnd(task: Task, report: Report, start: number) {
+        task.lastUpdate = new Date();
+        report.duration = Date.now() - start;
+        task.report = JSON.stringify(report, null, 4);
+        if (report.nbError > 0) {
+            task.status = -1;
+            const tab: string = '- ';
+            task.message = 'download failed in ' + report.duration + 'ms.\n'
+                + tab + report.nbSuccess + ' datasets were successfully downloaded\n'
+                + tab + report.nbError + ' datasets are (at least partially) in error and files could be missing.\n';
+            JSON.stringify(report);
+        } else {
+            task.status = task.status == -1 ? -1 : 1;
+            task.message = 'download completed in ' + report.duration + 'ms, ' + report.nbSuccess + ' files saved in the selected directory';
+        }
+
+        this.notificationService.pushLocalTask(task);
+    }
+
+    private recursiveSave(id: number, format: Format, userFolderHandle: FileSystemDirectoryHandle, remainingIds: number[], report: Report, task: Task, datasets?: Dataset[]): Promise<void> {
+        if (!id) return Promise.resolve();
+        return this.saveDataset(id, format, userFolderHandle, report, task, datasets?.find(ds => ds.id == id)).then(() => {
+            if (remainingIds.length > 0) {
+                return this.recursiveSave(remainingIds.shift(), format, userFolderHandle, remainingIds, report, task, datasets);
+            } else {
+                return Promise.resolve();
+            }
+        });
+    }
+
+    private saveDataset(id: number, format: Format, userFolderHandle: FileSystemDirectoryHandle, report: Report, task: Task, dataset?: Dataset): Promise<void> {
+        const metadataPromise: Promise<Dataset> = (dataset?.id == id && dataset.datasetAcquisition?.examination?.subject) ? Promise.resolve(dataset) : this.datasetService.get(id, 'lazy');
+        const downloadPromise: Promise<HttpResponse<Blob>> = this.datasetService.downloadToBlob(id, format);
```

Comment:
So at the end of the day, we download the datasets one by one ?
- We have a logic when there are multiple datasets (grouped by subject / exam for example) that may not be the same as in massive download ?
- In terms of efficiency, we load single dataset one by one, is it really more efficient ? (I need to test this !)